src/IO/HDF5/HDF5IOHandler.cpp

```diff
@@ -409,6 +436,46 @@ void HDF5IOHandlerImpl::createDataset(
             "[HDF5] Internal error: Failed to open HDF5 group during dataset "
             "creation");
 
+        if (m_handler->m_backendAccess == Access::APPEND)
+        {
+            // The dataset might already exist in the file from a previous run
+            // We delete it, otherwise we could not create it again with
+            // possibly different parameters.
+            // This is inefficient, as only the link to the dataset will be
+            // removed, but not the actual dataset
```

Comment:
Is it inefficient? I don't really know about HDF5's garbage collection mechanics. Do they need to be triggered manually? Do they apply in this situation?

Reply:
I think they are unlinked and not removed unless one does a HDF5 repack. I think that's also the right thing to do.

maybe @jeanbez knows is the below code looks solid to remove a HDF5 object? :)

Reply:
> I think that's also the right thing to do.

So you mean we should repack here?

Reply:
No, repack is an expensive operation one does in post. It copies everything.

Reply:
@ax3l I was not sure, so I checked with Tang/Suren and they mentioned that when removing an object with H5Ldelete, the HDF5 library should be able to detect and recycle the file space when no other reference to the deleted object exists. Regarding the code below it's more efficient to do H5Lexists to test whether a group/dset/attr exists, than iterating over existing items in a group like the one in this code snippet.

Reply:
Thanks a lot for the details, @jeanbez.

Can we also simply unlink a group and HDF5 garbage collects all underlying inodes of that group? (Similar for data sets?)

Reply:
The latest commit [7afb99d](https://github.com/openPMD/openPMD-api/pull/1007/commits/7afb99dbb71bde944a42969db99122732524222f) now uses H5Lexists, thanks for the tip!

Reply:
@ax3l Yes, according to the HDF5 team the refcount will be decremented on the group when it's unlinked from its parent, and the group will be deleted if its refcount drops to zero.  If a group is deleted, it will unlink all of its children, possibly causing them to be deleted as well.